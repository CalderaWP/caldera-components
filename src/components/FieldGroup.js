import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup,FormControl,ControlLabel,HelpBlock } from 'react-bootstrap';
export function FieldGroup({ id, label, help, ...props }) {
	return (
		<FormGroup controlId={id}>
			<ControlLabel
				controlId={id}
			>
				{label}
			</ControlLabel>
			<FormControl {...props} />
			{help && <HelpBlock>{help}</HelpBlock>}
		</FormGroup>
	);
}

FieldGroup.propTypes = {
	id: PropTypes.string.isRequired,
	props: PropTypes.object,
	help: PropTypes.string,
	label: PropTypes.string.isRequired
};