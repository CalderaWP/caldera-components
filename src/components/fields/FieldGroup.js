import React from 'react';
import classNames from 'classnames';
import {addOrRemoveFromArray, ariaDescribedbyAttr} from './util';
import {
	fieldGroupPropTypes,
} from './propTypes';

import {FieldInner} from './FieldInner';
import {RenderGroup} from '../RenderGroup';
import {Message} from './messages/Message';
import {messageObjectFactory} from './messages/messageObjectFactory';

/**
 * Represents one configField -- wrapper, label and input.
 *
 * @param props
 * @return {*}
 * @constructor
 */
export const FieldGroup = (props) => {

	/**
	 * Creates the id attribute
	 * @return {String}
	 */
	function idAttrForHelpElement() {
		return ariaDescribedbyAttr(props.id, props.help);
	}

	function checkboxChangeHandler(option,values) {
		if (!Array.isArray(values)) {
			if ('string' === typeof values || 'number' === typeof  values) {
				values = [values];
			} else {
				values = [];

			}
		}
		values = addOrRemoveFromArray(option.value, values);
		props.onValueChange(values);
	}

	/**
	 * Based on type, render the field.
	 *
	 * @return {*}
	 */
	function fieldInner(fieldProps){
		const message = 'object' === typeof  fieldProps.message
			? messageObjectFactory(fieldProps.message)
			: messageObjectFactory({message:null, error: false });
		function getFieldClassName(conditionalClassNames) {
			return classNames(
				{
					required: fieldProps.isRequired,
					...conditionalClassNames
				}
			);
		}

		/**
		 * Is this option the checked value?
		 * @param {object}option
		 * @param {String|number|Array} currentValue
		 * @return {boolean}
		 */
		function isCheckedOption( option, currentValue){
			if( Array.isArray( currentValue ) ){
				return currentValue.includes(option.value);
			}
			return option.value === currentValue;
		}

		//Fieldsets are rendered recursively.
		if( 'fieldset' === fieldProps.type ){
			return (
				<fieldset
					id={props.id}
				>
					<legend>
						{props.label}
					</legend>
					{props.options.map(option =>  {


						//Call this same function, as a regular checkbox
						return (
							<React.Fragment
								key={ `${props.id}-${option.value}` }
							>
								{fieldInner({
									type: 'input',
									inputType: 'checkbox',
									id: `${props.id}-${option.value}`,
									value: isCheckedOption(option,props.value),
									fieldClassName: getFieldClassName({
										'fieldset-checkbox' : true,
									}),
									label: option.label,
									disabled:props.disabled,
									onValueChange: () => {
										checkboxChangeHandler(option,props.values);
									}
								})}
							</React.Fragment>
						);
					})}
				</fieldset>
			);
		}else if( 'hidden' === fieldProps.inputType ){
			return (
				FieldInner(
					{
						type: props.type,
						id: props.id,
						fieldClassName: getFieldClassName(),
						value:fieldProps.value,
						onValueChange:fieldProps.onValueChange,
						inputType: fieldProps.inputType,
						disabled:props.disabled,
					}
				)
			);
		}else{
			return (
				<div
					className={RenderGroup.classNames.fieldWrapper}
				>
					<label
						htmlFor={fieldProps.id}
					>
						{fieldProps.label}
					</label>
					<Message
						message={message}
					/>
					{FieldInner(
						{
							type: fieldProps.type,
							id: fieldProps.id,
							fieldClassName: getFieldClassName({
								'block-input': fieldProps.isBlockInput
							}),
							help:fieldProps.help,
							value:fieldProps.value,
							onValueChange:fieldProps.onValueChange,
							inputType: fieldProps.inputType,
							options: fieldProps.options,
							disabled:props.disabled,
						}
					)}
					{fieldProps.help &&
					<p
						id={`${idAttrForHelpElement()}`}
						className={'description'}
					>
						{fieldProps.help}
					</p>
					}
				</div>
			);
		}

	}

	/**
	 * Render field group
	 */
	return fieldInner(props);
};




FieldGroup.propTypes = fieldGroupPropTypes;

FieldGroup.defaultProps = {
	isBlockInput: true,
	isRequired: false,
	help: ''
};