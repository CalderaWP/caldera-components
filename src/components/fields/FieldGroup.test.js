import {FieldGroup} from './FieldGroup';
import renderer from 'react-test-renderer';
import React from 'react';
import {mount} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

describe('Field Group component', () => {
	describe('Field group component props', () => {
		it('Works with help text', () => {
			const component = renderer.create(
				<FieldGroup
					id={'control-20'}
					label={'Who'}
					help={'Who to say hi to'}
					onValueChange={() => {
					}}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});

		it('Help text is optional', () => {
			const component = renderer.create(
				<FieldGroup
					id={'control-21'}
					label={'Who'}
					onValueChange={() => {
					}}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});

		it('Can show an input ', () => {
			const component = renderer.create(
				<FieldGroup
					id={'control-22'}
					label={'Who'}
					type={'input'}
					onValueChange={() => {
					}}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});

		it('Can mark inner input as required', () => {
			const component = renderer.create(
				<FieldGroup
					id={'control-223'}
					label={'Who'}
					type={'input'}
					isRequired={true}
					onValueChange={() => {
					}}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});

		it('Can mark inner input as disabled', () => {
			const component = renderer.create(
				<FieldGroup
					id={'control-224'}
					label={'Who'}
					type={'input'}
					isRequired={true}
					onValueChange={() => {
					}}
					disabled={true}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});
	});

	describe('Field group component rendering', () => {


		describe('inner input', () => {
			it('Has inner input', () => {
				const wrapper = mount(<FieldGroup
					id={'test-control'}
					label={'Who'}
					type={'input'}
					onValueChange={() => {
					}}
				/>);

				expect(wrapper.find('#test-control input').length).toBeTruthy();
			});

			it('controls inner input  html5 type', () => {
				const wrapper = mount(<FieldGroup
					id={'test-control'}
					label={'Who'}
					type={'input'}
					onValueChange={() => {
					}}
					value={'Sivans'}
					inputType={'hidden'}
				/>);

				expect(wrapper.find('#test-control input').prop('type')).toEqual('hidden');
			});

			it('Can change inner input', () => {
				let setValue = '';

				const wrapper = mount(<FieldGroup
					id={'test-control'}
					label={'Who'}
					type={'input'}
					onValueChange={(newValue) => {
						setValue = newValue;
					}}
					value={'Sivans'}
				/>);

				wrapper.find('#test-control input').simulate('change', {target: {value: 'r3s'}});
				expect(setValue).toEqual('r3s');

			});

			it('Can disable inner input', () => {
				const wrapper = mount(<FieldGroup
					id={'control-225'}
					label={'Who'}
					type={'input'}
					isRequired={true}
					onValueChange={() => {
					}}
					disabled={true}
				/>);

				expect(wrapper.find('#control-225 input').prop('disabled')).toEqual(true);

			});
		});

		describe('Label of Input', () => {
			it('Has no label for hidden fields ', () => {
				const wrapper = mount(<FieldGroup
					id={'test-control'}
					label={'I will not be outputed'}
					type={'input'}
					inputType={'hidden'}
					onValueChange={() => {
					}}
					value={'Sivans'}
				/>);
				expect(wrapper.find('input')).toHaveLength(1);
				expect(wrapper.find('label')).toHaveLength(0);
			});

			it('Has has label for other input fields ', () => {
				const wrapper = mount(<FieldGroup
					id={'test-control'}
					label={'I will not be outputed'}
					type={'input'}
					onValueChange={() => {
					}}
					value={'Sivans'}
				/>);
				expect(wrapper.find('input')).toHaveLength(1);
				expect(wrapper.find('label')).toHaveLength(1);
			});

		});

	});

	describe('Using as a select field', () => {

		it('Outputs a select field', () => {
			const wrapper = mount(<FieldGroup
				id={'select'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {

				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}
			/>);
			expect(wrapper.find('select')).toHaveLength(1);
			expect(wrapper.find('label')).toHaveLength(1);
		});

		it('Outputs a select field label', () => {
			const wrapper = mount(<FieldGroup
				id={'select2'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {

				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}
			/>);
			expect(wrapper.find('label')).toHaveLength(1);
		});

		it('Outputs the right select field label', () => {
			const wrapper = mount(<FieldGroup
				id={'select3'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {
				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}
			/>);
			expect(wrapper.find('label').text()).toEqual('Basic select field');
		});

		it('Can disable inner select', () => {
			const wrapper = mount(<FieldGroup
				id={'select3-b'}
				label={'Who'}
				type={'input'}
				isRequired={true}
				onValueChange={() => {
				}}
				disabled={true}
			/>);

			expect(wrapper.find('#select3-b input').prop('disabled')).toEqual(true);

		});
	});

	describe('<fieldset>', () => {
		describe('checkbox fieldsets', () => {
			it('Has a fieldset wrapper', () => {
				const wrapper = mount(
					<FieldGroup
						id={'fieldset-1'}
						label={'How many'}
						onValueChange={() => {
						}}
						inputType={'checkbox'}
						type={'fieldset'}
						options={[
							{
								value: 1,
								label: 'One'
							},
							{
								value: 2,
								label: 'Two'
							}
						]}
					/>
				);
				expect(wrapper.find('fieldset')).toHaveLength(1);
			});

			it('Has a legend inside of the fieldset ', () => {
				const wrapper = mount(
					<FieldGroup
						id={'fieldset-2'}
						label={'How many'}
						onValueChange={() => {
						}}
						inputType={'checkbox'}
						type={'fieldset'}
						options={[
							{
								value: 1,
								label: 'One'
							},
							{
								value: 2,
								label: 'Two'
							}
						]}
					/>
				);
				expect(wrapper.find('fieldset').children().find('legend')).toHaveLength(1);
			});

			it('Has a the right legend inside of the fieldset ', () => {
				const wrapper = mount(
					<FieldGroup
						id={'fieldset-3'}
						label={'How many'}
						onValueChange={() => {
						}}
						inputType={'checkbox'}
						type={'fieldset'}
						options={[
							{
								value: 1,
								label: 'One'
							},
							{
								value: 2,
								label: 'Two'
							}
						]}
					/>
				);
				expect(wrapper.find('fieldset').children().find('legend').text()).toEqual('How many');
			});

			it('Has 2 checkboxes inside of the fieldset', () => {
				const wrapper = mount(
					<FieldGroup
						id={'fieldset-3'}
						label={'How many'}
						onValueChange={() => {
						}}
						inputType={'checkbox'}
						type={'fieldset'}
						options={[
							{
								value: 1,
								label: 'One'
							},
							{
								value: 2,
								label: 'Two'
							}
						]}
					/>
				);
				expect(wrapper.find('fieldset').children().find('input')).toHaveLength(2);
			});

			it('Checkboxes have onValueChange prop', () => {
				const wrapper = mount(
					<FieldGroup
						id={'fieldset-3'}
						label={'How many'}
						onValueChange={() => {
						}}
						inputType={'checkbox'}
						type={'fieldset'}
						options={[
							{
								value: 1,
								label: 'One'
							},
							{
								value: 2,
								label: 'Two'
							}
						]}
					/>
				);
				expect(typeof wrapper.find('fieldset').children().find('input').first().prop( 'onChange' )).toEqual('function');
			});

		});

	});

	describe('Message group in fieldgroup', () => {
		it('Renders error with message', () => {
			const wrapper = mount(<FieldGroup
				id={'select-error-test'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {

				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}
				message={{
					message: 'Hi Roy',
					error: true,
				}}

			/>);
			expect(wrapper.find('.caldera-components-message' )).toHaveLength(1);
		});

		it('Renders non-error with message', () => {
			const wrapper = mount(<FieldGroup
				id={'select-error-test'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {

				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}
				message={{
					message: 'Hi Roy',
					error: false,
				}}

			/>);
			expect(wrapper.find('.caldera-components-message')).toHaveLength(1);
		});

		it('Renders the right error message', () => {
			const message = 'Hi Roy';
			const wrapper = mount(<FieldGroup
				id={'select-error-test-2'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {

				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}
				message={{
					message: message,
					error: true,
				}}

			/>);
			expect(wrapper.find('.caldera-components-message').text()).toBe(message);
		});

		it('Renders the right non-error message', () => {
			const message = 'Hi Roy';
			const wrapper = mount(<FieldGroup
				id={'select-error-test-3'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {

				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}
				message={{
					message: message,
					error: false,
				}}

			/>);
			expect(wrapper.find('.caldera-components-message').text()).toBe(message);
		});

		it('Renders nothing when message prop not passed', () => {
			const wrapper = mount(<FieldGroup
				id={'select-error-test-3b'}
				label={'Basic select field'}
				value={'1'}
				type={'select'}
				onValueChange={() => {

				}}
				options={[
					{
						value: 1,
						label: 'One'
					},
					{
						value: 2,
						label: 'Two'
					}
				]}

			/>);
			expect(wrapper.find('.caldera-components-message' )).toHaveLength(0);
		});

		it('Renders nothing when message prop passed, without a message', () => {
			const wrapper = mount(
				<FieldGroup
					id={'select-error-test-4'}
					label={'Basic select field'}
					value={'1'}
					type={'select'}
					onValueChange={() => {

					}}
					options={[
						{
							value: 1,
							label: 'One'
						},
						{
							value: 2,
							label: 'Two'
						}
					]}
					message={{
						error: true,
					}}

				/>
			);
			expect(wrapper.find('.caldera-components-message')).toHaveLength(0);
		});
	});

	it( 'Uses the Magic group, if type is magic', () => {
		const component = renderer.create(
			<FieldGroup
				id={'magic-field-group'}
				label={'magic'}
				type={'magic'}
				onValueChange={() => {}}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
});