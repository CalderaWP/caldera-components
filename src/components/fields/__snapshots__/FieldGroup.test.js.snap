// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Group component Field group component props Can mark inner input as disabled 1`] = `
<div
  className="caldera-config-field"
>
  <label
    htmlFor="control-224"
  >
    Who
  </label>
  <input
    aria-describedby={null}
    className="required block-input field-config field-config"
    disabled={true}
    id="control-224"
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    required={true}
    type="text"
    value={undefined}
  />
  
</div>
`;

exports[`Field Group component Field group component props Can mark inner input as required 1`] = `
<div
  className="caldera-config-field"
>
  <label
    htmlFor="control-223"
  >
    Who
  </label>
  <input
    aria-describedby={null}
    className="required block-input field-config field-config"
    disabled={undefined}
    id="control-223"
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    required={true}
    type="text"
    value={undefined}
  />
  
</div>
`;

exports[`Field Group component Field group component props Can show an input  1`] = `
<div
  className="caldera-config-field"
>
  <label
    htmlFor="control-22"
  >
    Who
  </label>
  <input
    aria-describedby={null}
    className="block-input field-config field-config"
    disabled={undefined}
    id="control-22"
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    required={false}
    type="text"
    value={undefined}
  />
  
</div>
`;

exports[`Field Group component Field group component props Help text is optional 1`] = `
<div
  className="caldera-config-field"
>
  <label
    htmlFor="control-21"
  >
    Who
  </label>
  <input
    aria-describedby={null}
    className="block-input field-config field-config"
    disabled={undefined}
    id="control-21"
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    required={false}
    type="text"
    value={undefined}
  />
  
</div>
`;

exports[`Field Group component Field group component props Works with help text 1`] = `
<div
  className="caldera-config-field"
>
  <label
    htmlFor="control-20"
  >
    Who
  </label>
  <input
    aria-describedby="control-20-description"
    className="block-input field-config field-config"
    disabled={undefined}
    id="control-20"
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    required={false}
    type="text"
    value={undefined}
  />
  <p
    className="description"
    id="control-20-description"
  >
    Who to say hi to
  </p>
</div>
`;

exports[`Field Group component Uses the Magic group, if type is magic 1`] = `
<div
  className="caldera-magic-select-group caldera-config-field"
>
  <label
    htmlFor="magic-field-group"
  >
    magic
  </label>
  
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <input
      aria-autocomplete="list"
      aria-expanded={false}
      autoComplete="off"
      className="caldera-magic-input caldera-magic-input-magic-field-group"
      id="magic-field-group"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="combobox"
      value=""
    />
  </div>
</div>
`;
